libraries/AP_HAL_Linux/Perf.cpp

```diff
@@ -186,4 +195,322 @@ void Util::perf_count(perf_counter_t perf)
+        hal.console->printf("Missing implementation of append_perf_data() "
+                            "for PC_INTERVAL type.\n");
+        return 0;
+    }
+
+    if (perf_base->type == Util::PC_COUNT) {
+        perf_counter_count_t *perf_count = (perf_counter_count_t *)perf;
+        return sd_bus_message_append(reply, "{st}", "count", perf_count->count);
+    }
+
+    perf_counter_elapsed_t *perf_elapsed = (perf_counter_elapsed_t *)perf;
+
+    int ret = sd_bus_message_append(reply, "{st}", "count",
+                                    perf_elapsed->count);
+    if (ret < 0) {
+        return ret;
+    }
+
+    ret = sd_bus_message_append(reply, "{st}", "total", perf_elapsed->total);
+    if (ret < 0) {
+        return ret;
+    }
+
+    uint64_t temp = perf_elapsed->count;
+    if (temp == 0) {
+        temp = 1;
+    }
+    temp = perf_elapsed->total / temp;
+    ret = sd_bus_message_append(reply, "{st}", "average", temp);
+    if (ret < 0) {
+        return ret;
+    }
+
+    ret = sd_bus_message_append(reply, "{st}", "minimum", perf_elapsed->least);
+    if (ret < 0) {
+        return ret;
+    }
+
+    ret = sd_bus_message_append(reply, "{st}", "maximum", perf_elapsed->most);
+    if (ret < 0) {
+        return ret;
+    }
+
+    double variance;
+    if (perf_elapsed->count > 1) {
+        variance = perf_elapsed->m2 / (perf_elapsed->count - 1);
+    } else {
+        variance = 0.0;
+    }
+    temp = round(variance);
+    ret = sd_bus_message_append(reply, "{st}", "variance", temp);
+    if (ret < 0) {
+        return ret;
+    }
+
+    temp = round(sqrt(variance));
+    return sd_bus_message_append(reply, "{st}", "std-deviation", temp);
+}
+
+static int append_perf(perf_counter_base_t *perf, sd_bus_message *reply)
+{
+    int r = sd_bus_message_open_container(reply, SD_BUS_TYPE_STRUCT, "ssa{st}");
+    if (r < 0) {
+        return r;
+    }
+
+    r = sd_bus_message_append(reply, "ss", perf->name, get_perf_type(perf));
+    if (r < 0) {
+        return r;
+    }
+
+    r = sd_bus_message_open_container(reply, 'a', "{st}");
+    if (r < 0) {
+        return r;
+    }
+
+    r = append_perf_specific_data(perf, reply);
+    if (r < 0) {
+        return r;
+    }
+
+    /* closing a{st} */
+    r = sd_bus_message_close_container(reply);
+    if (r < 0) {
+        return r;
+    }
+
+    /* closing (ssa{st}) */
+    return sd_bus_message_close_container(reply);
+}
+
+int PerfManager::dbus_get_perf_callback(sd_bus_message *m, void *userdata, sd_bus_error *ret_error)
+{
+    int r = sd_bus_message_enter_container(m, SD_BUS_TYPE_ARRAY, "s");
+    if (r < 0) {
+        return sd_bus_error_set_errno(ret_error, r);
+    }
+
+    std::vector<const char *> names;
+    while (true) {
+        const char *name;
+        /* return 0 when there is no item left on array */
+        r = sd_bus_message_read_basic(m, SD_BUS_TYPE_STRING, &name);
+        if (r < 1) {
+            break;
+        }
+        names.push_back(name);
```

Comment:
here you will realloc at each iteration. Don't do this. std::vector should only be used together with reserve()

However you are only using it to iterate over the perf events. I think it would be better to just go and try... i.e. move this sd_bus_message_read_basic() to inside the loop. And fail gracefully if any error occurs.